Add partial copy constructors for TriggerActivity and TriggerCandidate

JamesJieranShen opened the pull request:
These constructors allow `TriggerActivity` and `TriggerCandidate` to be directly created from their corresponding data structs in `trgdataformat`.

Motivation for this is due to the fact that TAs are stored offline as `trgdataformat::TriggerActivityData` objects, yet the TC algorithms expects to see `triggeralgs::TriggerActivity` objects. There's no obvious way to convert between the two classes other than static_cast wizardries that may lead to unsafe memory operations.

JamesJieranShen committed:
Add partial copy constructors for TriggaerActivity and TriggerCandidate

Also explicitly define rule-of-five ctors for both classes.

MRiganSUSX commented:
Hi @JamesJieranShen,
This seems good to me. 

I was wondering whether we should be fully explicit and follow the rule of 5, so also adding the rest, as defaults. Something like:
```
  TriggerActivity(const TriggerActivity&) = default;
  TriggerActivity(TriggerActivity&&) = default;
  TriggerActivity& operator=(const TriggerActivity&) = default;
  TriggerActivity& operator=(TriggerActivity&&) = default;
  ~TriggerActivity() = default;
```

Thoughts?

JamesJieranShen commented:
Hey @MRiganSUSX ,

Thanks for the suggestions. I do think the rest of the constructors _should_ be automatically generated still, but I've added them explicitly to be safe, like you suggested.

MRiganSUSX commented:
Thanks for the updates @JamesJieranShen.
Seems like you missed the default move constructor in both cases? 
`TriggerCandidate(TriggerCandidate&&) = default;`

ArturSztuc reviewed (commented):
@aeoranday raised concerns as to how this PR is a "fix" that avoids solving the real problem, which I agree with.

In principle TCMaker should not be able to accept `TriggerActivity`, and only accept `TriggerActivityData`, and MLT should be able to accept `TriggerCandidateData` rather than `TriggerCandidate`.

The proper solution to the problem would be to change the TCMakers to only accept TriggerActivityData, and MLT to only accept TriggerCandidateData.

I think in principle we could accept this PR, but I think only as a "temporary fix" to be removed sooner than later, and only if changing TriggerCandidates/MLT is not doable on short-ish timescale.

JamesJieranShen commented:
Howdy, I've addressed the comments from @MRiganSUSX  -- sorry that I missed that earlier.

I think I agree with what @ArturSztuc suggested. TriggerActivity and TriggerCandidate _should_ stay internal to triggeralgs.

MRiganSUSX reviewed (approved):
From my perspective:
- code looks good
- passes unittests
- passes integtest bundle
- passes trigger related systests (replay, bitwords)

I have not tested how this works with dunetrigger.

I'm fine with getting this merged, but I will leave the final decision to @ArturSztuc. If you decide to merge this, please also open an issue on the other changes required to makers/mlt.

ArturSztuc commented:
I tagged this for `v5.0.0` release. We need to have a bit wider discussion on this, perhaps at the next DS/PP meeting.

alessandrothea reviewed (approved):
LGTM
